### PR TITLE
Cadidate selection check assignment

### PR DESCRIPTION
**Pull request by montekki**
> This subsystem is only ever interested in parablocks assigned to the particular parachain which this validator is currently handling.

### CODE REVIEW

File: node/core/candidate-selection/src/lib.rs [hunk 1]
```diff
@@ -206,6 +281,16 @@ impl CandidateSelectionJob {
...
 		let _timer = self.metrics.time_handle_collation();
 
+		if self.assignment != para_id {
+			tracing::info!(
+				target: LOG_TARGET,
+				"Collator {:?} sent a collation outside of our assignment {:?}",
+				collator_id,
+				para_id,
+			);
+			return;
```
Comment (montekki) [inline node/core/candidate-selection/src/lib.rs:291]: probably we should report or disconnect from this collator

File: node/core/candidate-selection/src/lib.rs [hunk 2]
```diff
@@ -19,27 +19,38 @@
 
 #![deny(missing_docs, unused_crate_dependencies, unused_results)]
 
+use std::collections::HashMap;
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:22]:
```suggestion
```

File: node/core/candidate-selection/src/lib.rs [hunk 3]
```diff
@@ -19,27 +19,38 @@
...
+	request_from_runtime,
+	Validator,
+	JobTrait, ToJobTrait, FromJobCommand,
 	metrics::{self, prometheus},
 };
-use polkadot_primitives::v1::{CandidateReceipt, CollatorId, Hash, Id as ParaId, PoV};
+use polkadot_primitives::v1::{
+	CandidateReceipt, CollatorId, CoreState, CoreIndex, Hash, Id as ParaId, PoV,
+};
 use std::{convert::TryFrom, pin::Pin};
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:47]:
```suggestion
use std::{convert::TryFrom, pin::Pin, collections::HashMap};
```

File: node/core/candidate-selection/src/lib.rs [hunk 4]
```diff
@@ -126,24 +137,86 @@ impl JobTrait for CandidateSelectionJob {
...
+		relay_parent: Hash,
+		keystore: Self::RunArgs,
 		metrics: Self::Metrics,
 		receiver: mpsc::Receiver<ToJob>,
-		sender: mpsc::Sender<FromJob>,
+		mut sender: mpsc::Sender<FromJob>,
 	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
 		Box::pin(async move {
-			let job = CandidateSelectionJob::new(metrics, sender, receiver);
+			macro_rules! try_runtime_api {
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:157]: Does this macro really needs to be declared in this function?

File: node/core/candidate-selection/src/lib.rs [hunk 5]
```diff
@@ -126,24 +137,86 @@ impl JobTrait for CandidateSelectionJob {
...
+							// requisite runtime API data. But these errors should not take
+							// down the node.
+							return Ok(());
+						}
+					}
+				}
+			}
+
+			let (groups, cores) = futures::try_join!(
+				request_validator_groups(relay_parent, &mut sender).await?,
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:178]: Below we just print a warning, but here we return the error?

File: node/core/candidate-selection/src/lib.rs [hunk 6]
```diff
@@ -126,24 +137,86 @@ impl JobTrait for CandidateSelectionJob {
...
+
+			let mut assignment = None;
+			for (idx, core) in cores.into_iter().enumerate() {
+				// Ignore prospective assignments on occupied cores for the time being.
+				if let CoreState::Scheduled(scheduled) = core {
+					let core_index = CoreIndex(idx as _);
+					let group_index = group_rotation_info.group_for_core(core_index, n_cores);
+					if let Some(g) = validator_groups.get(group_index.0 as usize) {
+						if g.contains(&validator.index()) {
+							assignment = Some(scheduled.para_id);
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:207]:
```suggestion
							assignment = Some(scheduled.para_id);
							break;
```

?

File: node/core/candidate-selection/src/lib.rs [hunk 7]
```diff
@@ -126,24 +137,86 @@ impl JobTrait for CandidateSelectionJob {
...
+						}
+						let _ = groups.insert(scheduled.para_id, g.clone());
```
Comment (bkchr) [inline node/core/candidate-selection/src/lib.rs:209]: What is the purpose of this?

File: node/core/candidate-selection/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply node/core/candidate-selection/src/lib.rs:291]: Yeah. Collators shouldn't be advertising to validators who aren't in their group.

File: node/core/candidate-selection/src/lib.rs (hunk 7 above)
Comment (coriolinus) [reply node/core/candidate-selection/src/lib.rs:209]: Looks like it's related to `deny(unused_results)` at the top of the file.

Comment (bkchr) [reply node/core/candidate-selection/src/lib.rs:209]: That was not my question. `groups` is not used for anything.

And this lint is useless anyway...

File: node/core/candidate-selection/src/lib.rs (hunk 4 above)
Comment (bkchr) [reply node/core/candidate-selection/src/lib.rs:157]: Why? This blows up the function and makes it unreadable. Putting it above will not put it into any broader scope, besides the scope of this module.

If we follow this reasoning, every free standing function that is called in here should be declared in the function that uses it.

File: node/core/candidate-selection/src/lib.rs (hunk 7 above)
Comment (montekki) [reply node/core/candidate-selection/src/lib.rs]: removed

File: node/core/candidate-selection/src/lib.rs (hunk 4 above)
Comment (montekki) [reply node/core/candidate-selection/src/lib.rs]: moved outside

File: node/core/candidate-selection/src/lib.rs [hunk 8]
```diff
@@ -57,30 +63,92 @@ enum Error {
...
 	type ToJob = CandidateSelectionMessage;
 	type Error = Error;
-	type RunArgs = ();
+	type RunArgs = SyncCryptoStorePtr;
 	type Metrics = Metrics;
 
 	const NAME: &'static str = "CandidateSelectionJob";
 
-	#[tracing::instrument(skip(_relay_parent, _run_args, metrics, receiver, sender), fields(subsystem = LOG_TARGET))]
+	#[tracing::instrument(skip(relay_parent, keystore, metrics, receiver, sender), fields(subsystem = LOG_TARGET))]
```
Comment (montekki) [inline node/core/candidate-selection/src/lib.rs:94]:
```suggestion
	#[tracing::instrument(skip(keystore, metrics, receiver, sender), fields(subsystem = LOG_TARGET))]
```

**Review by rphmeier — APPROVED** [review]

**Pull request merged**